### PR TITLE
Refactor processing of common terms in phrases

### PR DESCRIPTION
**Pull request by horpto**

### CODE REVIEW

**Comment by horpto** [comment]
review, please

**Comment by piskvorky** [comment]
This kind of low-level looping is hard to review from code… can you extend the unit tests first @horpto ?

If you could pick out some **nasty corner cases** for the tests (empty inputs, phrases first/last position, scores that equal threshold at exact float equality, etc) that'd be awesome (and also useful beyond this refactoring).

File: gensim/models/phrases.py [hunk 1]
```diff
@@ -154,35 +153,33 @@ def analyze_sentence(self, sentence, threshold, common_terms, scorer):
         # adding None is a trick that helps getting an automatic happy ending
         # as it won't be a common_word, nor score
         s.append(None)
-        last_uncommon = None
-        in_between = []
+        common = []
```
[the remark this answers is not in the document]
Comment (mpenkov) [reply gensim/models/phrases.py]: Removing the last_uncommon variable invalidates the comment on L161.

File: gensim/models/phrases.py [hunk 2]
```diff
@@ -482,19 +479,17 @@ def learn_vocab(sentences, max_vocab_size, delimiter=b'_', progress_per=10000,
...
-                    in_between = []
-                elif last_uncommon is not None:
-                    in_between.append(word)
+                    if common:
+                        common.append(word)
+                        vocab[delimiter.join(common)] += 1
+                    common = [word]
+                elif common:
+                    common.append(word)
                 total_words += 1
```
Comment (mpenkov) [inline gensim/models/phrases.py]: We don't need to increment this every single time, do we?

```suggestion
            total_words += len(sentence)
```

File: gensim/models/phrases.py [hunk 3]
```diff
@@ -482,19 +479,17 @@ def learn_vocab(sentences, max_vocab_size, delimiter=b'_', progress_per=10000,
                     "PROGRESS: at sentence #%i, processed %i words and %i word types",
                     sentence_no, total_words, len(vocab),
                 )
-            s = [utils.any2utf8(w) for w in sentence]
-            last_uncommon = None
-            in_between = []
-            for word in s:
+            common = []
```
[the remark this answers is not in the document]
Comment (mpenkov) [reply gensim/models/phrases.py]: Also, `common` isn't a particularly good variable name here, because it clashes with `common_terms` and means something completely different.

I think something like `bigram` would be a better name.

**Comment by mpenkov** [comment]
Interestingly, at least part of the updated code isn't covered by tests **at all**, so we should really add some good tests before refactoring:

```
(gensim) misha@cabron:~/git/gensim$ grep -R learn_vocab gensim
Binary file gensim/models/__pycache__/phrases.cpython-37.pyc matches
Binary file gensim/models/__pycache__/phrases.cpython-35.pyc matches
gensim/models/phrases.py:    def learn_vocab(sentences, max_vocab_size, delimiter=b'_', progress_per=10000,
gensim/models/phrases.py:            >>> pruned_words, counters, total_words = Phrases.learn_vocab(sentences, 100)
gensim/models/phrases.py:        # counts collected in previous learn_vocab runs.
gensim/models/phrases.py:        min_reduce, vocab, total_words = self.learn_vocab(
Binary file gensim/models/phrases.pyc matches
```

**Comment by mpenkov** [comment]
@horpto What is the status of this PR?
